src/plugins/data_source/server/saved_objects/data_source_permission_client_wrapper.ts

```diff
@@ -0,0 +1,173 @@
+      attributes: Partial<T>,
+      options: SavedObjectsUpdateOptions = {}
+    ): Promise<SavedObjectsUpdateResponse<T>> => {
+      if (this.hasNoDataSourcePermission(type)) {
+        throw this.generatePermissionError();
+      }
+      return await wrapperOptions.client.update(type, id, attributes, options);
+    };
+
+    const bulkUpdateWithEditMode = async <T = unknown>(
+      objects: Array<SavedObjectsBulkUpdateObject<T>>,
+      options?: SavedObjectsBulkUpdateOptions
+    ): Promise<SavedObjectsBulkUpdateResponse<T>> => {
+      const disallowedSavedObjects: Array<SavedObjectsBulkUpdateObject<T>> = [];
+      const allowedSavedObjects: Array<SavedObjectsBulkUpdateObject<T>> = [];
+
+      objects.forEach((item) => {
+        if (this.hasNoDataSourcePermission(item.type)) {
+          disallowedSavedObjects.push(item);
+          return;
+        }
+
+        allowedSavedObjects.push(item);
+        return;
+      });
+
+      const bulkUpdateResult = await wrapperOptions.client.bulkUpdate(allowedSavedObjects, options);
+
+      // Merge the data source saved objects and real client bulkUpdate result.
+      return {
+        saved_objects: [
+          ...(bulkUpdateResult?.saved_objects || []),
+          ...disallowedSavedObjects.map((item) => ({
+            ...item,
+            error: {
+              ...this.generatePermissionError().output.payload,
+              metadata: { isNotOverwritable: true },
+            },
+          })),
+        ],
+      } as SavedObjectsBulkUpdateResponse<T>;
+    };
+
+    const deleteWithEditMode = async (
+      type: string,
+      id: string,
+      options: SavedObjectsDeleteOptions = {}
+    ) => {
+      if (this.hasNoDataSourcePermission(type)) {
+        throw this.generatePermissionError();
+      }
+      return await wrapperOptions.client.delete(type, id, options);
+    };
+
+    const isDashboardAdmin = getWorkspaceState(wrapperOptions.request)?.isDashboardAdmin;
+    if (this.editMode === EditMode.AdminOnly && isDashboardAdmin) {
+      return wrapperOptions.client;
+    }
+
+    return {
+      ...wrapperOptions.client,
+      create: createWithEditMode,
+      bulkCreate: bulkCreateWithEditMode,
+      checkConflicts: wrapperOptions.client.checkConflicts,
+      delete: deleteWithEditMode,
+      find: wrapperOptions.client.find,
+      bulkGet: wrapperOptions.client.bulkGet,
+      get: wrapperOptions.client.get,
+      update: updateWithEditMode,
+      bulkUpdate: bulkUpdateWithEditMode,
+      errors: wrapperOptions.client.errors,
+      addToNamespaces: wrapperOptions.client.addToNamespaces,
+      deleteFromNamespaces: wrapperOptions.client.deleteFromNamespaces,
+    };
+  };
+
+  constructor(private editMode: string) {}
+
+  private hasNoDataSourcePermission = (type: string): boolean => {
+    if (DATA_SOURCE_SAVED_OBJECT_TYPE !== type) {
+      return false;
+    }
+
+    return this.editMode === EditMode.ReadOnly || this.editMode === EditMode.AdminOnly;
```

Comment:
Since this client wrapper only be attached after editMode is not `None` in [this line](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7091/files#diff-a01a6007a1a466a29810fbf0689bc607c3c3dfc9d3f7713da9bbf653beb92ef8R77), do we need to check the editMode here?

Reply:
For the readability of the code, I did not directly `return true` here.